kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,759 @@
+          .build())
+    }
+    fileSpecBuilder.addType(adapter)
+  }
+}
+
+private fun ProtoBuf.TypeParameter.asTypeName(
+    nameResolver: NameResolver,
+    getTypeParameter: (index: Int) -> ProtoBuf.TypeParameter): TypeName {
+  return TypeVariableName(
+      name = nameResolver.getString(name),
+      bounds = *(upperBoundList.map { it.asTypeName(nameResolver, getTypeParameter) }
+          .toTypedArray()),
+      variance = variance.asKModifier()
+  )
+}
+
+private fun ProtoBuf.TypeParameter.Variance.asKModifier(): KModifier? {
+  return when (this) {
+    Variance.IN -> IN
+    Variance.OUT -> OUT
+    Variance.INV -> null
+  }
+}
+
+/**
+ * Returns the TypeName of this type as it would be seen in the source code,
+ * including nullability and generic type parameters.
+ *
+ * @param [nameResolver] a [NameResolver] instance from the source proto
+ * @param [getTypeParameter]
+ * A function that returns the type parameter for the given index.
+ * **Only called if [ProtoBuf.Type.hasTypeParameter] is `true`!**
+ */
+private fun ProtoBuf.Type.asTypeName(
+    nameResolver: NameResolver,
+    getTypeParameter: (index: Int) -> ProtoBuf.TypeParameter
+): TypeName {
+
+  val argumentList = when {
+    hasAbbreviatedType() -> abbreviatedType.argumentList
+    else -> argumentList
+  }
+
+  if (hasFlexibleUpperBound()) {
+    return WildcardTypeName.subtypeOf(
+        flexibleUpperBound.asTypeName(nameResolver, getTypeParameter))
+  } else if (hasOuterType()) {
+    return WildcardTypeName.supertypeOf(outerType.asTypeName(nameResolver, getTypeParameter))
+  }
+
+  val realType = when {
+    hasTypeParameter() -> return getTypeParameter(typeParameter)
+        .asTypeName(nameResolver, getTypeParameter)
+    hasTypeParameterName() -> typeParameterName
+    hasAbbreviatedType() -> abbreviatedType.typeAliasName
+    else -> className
+  }
+
+  var typeName: TypeName = ClassName.bestGuess(nameResolver.getString(realType)
+      .replace("/", "."))
+
+  if (argumentList.isNotEmpty()) {
+    val remappedArgs: Array<TypeName> = argumentList.map {
+      val projection = if (it.hasProjection()) {
+        it.projection
+      } else null
+      if (it.hasType()) {
+        it.type.asTypeName(nameResolver, getTypeParameter)
+            .let { typeName ->
+              projection?.let {
+                when (it) {
+                  Projection.IN -> WildcardTypeName.supertypeOf(typeName)
+                  Projection.OUT -> {
+                    if (typeName == ANY) {
+                      // This becomes a *, which we actually don't want here.
+                      // List<Any> works with List<*>, but List<*> doesn't work with List<Any>
+                      typeName
+                    } else {
+                      WildcardTypeName.subtypeOf(typeName)
+                    }
+                  }
+                  Projection.STAR -> WildcardTypeName.subtypeOf(ANY)
+                  Projection.INV -> TODO("INV projection is unsupported")
```

Comment:
Is this an actual TODO or should it be `error(..)`?

Reply:
I actually don't know. I forgot about that and meant to ask what you want to do with this case